src/pages/home/report/ReportActionContextMenu.js

```diff
@@ -92,29 +30,93 @@ const propTypes = {
 };
 
 const defaultProps = {
-    reportAction: {},
     isMini: false,
     isVisible: false,
 };
 
-const ReportActionContextMenu = (props) => {
-    const wrapperStyle = getReportActionContextMenuStyles(props.isMini);
-    return props.isVisible && (
-        <View style={wrapperStyle}>
-            {CONTEXT_ACTIONS.map(contextAction => contextAction.shouldShow && (
-                <ReportActionContextMenuItem
-                    icon={contextAction.icon}
-                    text={contextAction.text}
-                    successIcon={contextAction.successIcon}
-                    successText={contextAction.successText}
-                    isMini={props.isMini}
-                    onPress={() => contextAction.onPress(props.reportAction)}
-                    key={contextAction.text}
-                />
-            ))}
-        </View>
-    );
-};
+class ReportActionContextMenu extends React.Component {
+    constructor(props) {
+        super(props);
+
+        // A list of all the context actions in this menu.
+        this.CONTEXT_ACTIONS = [
+            // Copy to clipboard
+            {
+                text: 'Copy to Clipboard',
+                icon: ClipboardIcon,
+                successText: 'Copied!',
+                successIcon: Checkmark,
+                shouldShow: true,
+
+                // If return value is true, we switch the `text` and `icon` on
+                // `ReportActionContextMenuItem` with `successText` and `successIcon` which will fallback to
+                // the `text` and `icon`
+                onPress: () => {
+                    const message = _.last(lodashGet(this.props.reportAction, 'message', null));
+                    const html = lodashGet(message, 'html', '');
+                    const text = lodashGet(message, 'text', '');
+                    const isAttachment = _.has(this.props.reportAction, 'isAttachment')
+                        ? this.props.reportAction.isAttachment
+                        : isReportMessageAttachment(text);
+                    if (!isAttachment) {
+                        Clipboard.setString(text);
+                    } else {
+                        Clipboard.setString(html);
+                    }
+                },
+            },
+
+            {
+                text: 'Copy Link',
+                icon: LinkCopy,
+                shouldShow: false,
+                onPress: () => {},
+            },
+
+            {
+                text: 'Mark as Unread',
+                icon: Mail,
+                successIcon: Checkmark,
+                shouldShow: true,
+                onPress: () => updateLastReadActionID(this.props.reportID, this.props.reportAction.sequenceNumber - 1),
```

Comment:
It's pretty confusing to see the `- 1` here, and then a `+ 1` to this value in `updateLastReadActionID()`. Can that be cleaned up at all so that all the math necessary is done inside the action method?

Reply:
I had quite the fight with this 😆 . Let's see if I can explain:

- There are two different things `updateLastReadActionID` does: set the last read sequence and place the `new` marker.
- When messages come in normally (and the chat is open), that new sequence is the last read sequence. All good there.
- However, when we manually mark the message as unread, the `last read sequence` is not the one we are marking as unread (that one is unread!). The last read sequence is the previous one, so that's what we are passing.
- The New marker, however, is displayed above the `first unread sequence`, not the `last read sequence`, hence the +1 over there.

I tried a few combinations of these, but I think this is what made the most sense, since we are passing the right sequence to `updateLastReadActionID` and retrieving the right sequence when setting the `new` marker.

Reply:
OK, I don't think it's very great that `updateLastReadActionID` is doing two different things (and its name is also now slightly misleading). Would separating out the logic into two different methods help? I think it might after looking at the code. Especially since there is only one place that uses the second param of `updateLastReadActionID()`.

The thing I don't like about having `- 1` here is that no one else trying to use the method `updateLastReadActionID()` would ever know that they need to subtract one from the sequence number.